### DISCUSSION TITLE
When may I need libusb on Android?

### DISCUSSION THREAD

**Discussion by SentretC** [libusb/libusb#1356] [Q&A] [answered]
In other words, what can be done with libusb that can't be done with the `android.hardware.usb` classes? (Besides utilizing existing code for other platforms)

I'm asking because I'm tasked to develop an app for our device which uses Cypress EZ-USB FX2LP for USB communication. The device works fine on Windows PC thanks to libusb, but on Android things are different:
- Some phones can recognize it
- Some phones power the device, but `usbManager.getDeviceList().size()` returns 0, however these phones can recognize other USB devices including non-general-purpose ones
- Some phones don't power the device and `usbManager.getDeviceList().size()` returns 0. These phones also don't recognize some other devices but do recognize flash drives

I'm wondering if more phones would work if I use libusb instead of the `android.hardware.usb` classes, but the readme says I still need to use those classes to request for permission on unrooted phones...

Also, any suggestions on how I can check where do things go wrong on those phones?

**Comment by SentretC** [comment]
Oh it seems that the problem is in our hardware design. We just bought a demo kit and it can be recognized on all phones we tested.
Still, I think answering the question can be helpful so I'm not closing it.

**Comment by Youw** [comment, marked as answer]
To answer you question someone has to go over the Android USB API and compare it with libusb API (and check what is actually implemented/functional on Android and in what cases).

I don't believe anyone ever did so.

My best _guess_ is that both should be functionally identical.

I'd make a few potential assumptions (didn't really check them):
- Android API _may_ have some _convenient_ Android-specific quirks, like integration with event loops or some other Android APIs;
- libusb API _may_ have an implementation for some things (maybe rarely used features) that might be ommited by Android API only because "no one needed them on Android side of things" and because libusb tries to implement all of the USB features there are in spec;

NOTE: when it comes to device enumeration/open - unless you're on rooted device, libusb (its user) still needs to use Android USB manager - that's the only way to get permission from the OS to open a Device.

---
UPD: and more straightforward answer:
If you're aming cross-platform implementation for you application - you definitely should consider libusb (at least for all "business logic").
If you're only targeting Android as an implementation for you application (or you're going to write an "entirely different application" for other platforms), specially if it is written in Java/Kotlin - Android USB API should be your obvious simple choice.
